File: src/main/java/com/itsci/mju/maebanjumpen/login/service/impl/LoginServiceImpl.java
package com.itsci.mju.maebanjumpen.login.service.impl;

import com.itsci.mju.maebanjumpen.login.dto.LoginDTO;
import com.itsci.mju.maebanjumpen.partyrole.dto.PartyRoleDTO;
import com.itsci.mju.maebanjumpen.exception.AccountStatusException; // ⬅️ เพิ่ม import
import com.itsci.mju.maebanjumpen.mapper.LoginMapper;
import com.itsci.mju.maebanjumpen.mapper.PartyRoleMapper;
import com.itsci.mju.maebanjumpen.entity.*;
import com.itsci.mju.maebanjumpen.login.repository.LoginRepository;
import com.itsci.mju.maebanjumpen.partyrole.repository.PartyRoleRepository;
import com.itsci.mju.maebanjumpen.person.repository.PersonRepository;
import com.itsci.mju.maebanjumpen.login.service.LoginService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class LoginServiceImpl implements LoginService {

    private final LoginRepository loginRepository;
    private final PersonRepository personRepository;
    private final PartyRoleRepository partyRoleRepository;
    private final LoginMapper loginMapper;
    private final PartyRoleMapper partyRoleMapper;

    @Override
    @Transactional(readOnly = true)
    public PartyRoleDTO authenticate(String username, String password) {
        return findPartyRoleByLogin(username, password);
    }

    @Override
    @Transactional
    public LoginDTO saveLogin(LoginDTO loginDto) {
        Login login = loginMapper.toEntity(loginDto);
        if (login.getPassword() == null) {
            throw new IllegalArgumentException("Password cannot be null");
        }
        login.setPassword(PasswordUtil.hashPassword(login.getPassword()));
        Login savedLogin = loginRepository.save(login);
        return loginMapper.toDto(savedLogin);
    }

    @Override
    @Transactional(readOnly = true)
    public LoginDTO getLoginByUsername(String username) {
        Optional<Login> loginOpt = loginRepository.findById(username);
        return loginOpt.map(loginMapper::toDto).orElse(null);
    }

    @Override
    @Transactional
    public void deleteLogin(String username) {
        loginRepository.deleteById(username);
    }

    @Override
    public LoginDTO updateLogin(String username, LoginDTO loginDto) {
        return null;
    }

    @Override
    @Transactional(readOnly = true)
    public PartyRoleDTO findPartyRoleByLogin(String username, String rawPassword) {
        Optional<Login> loginOpt = loginRepository.findByUsername(username);
        if (loginOpt.isEmpty()) {
            return null;
        }

        Login storedLogin = loginOpt.get();
        String storedHash = storedLogin.getPassword();

        if (!PasswordUtil.verifyPassword(rawPassword, storedHash)) {
            System.out.println("-> [LoginService] Authentication failed: Invalid password for user: " + username);
            return null;
        }

        Optional<Person> personOpt = personRepository.findByLoginUsername(username);
        if (personOpt.isEmpty()) {
            System.out.println("-> [LoginService] Authentication failed: Person not found for user: " + username);
            return null;
        }
        Person person = personOpt.get();

        String status = person.getAccountStatus();
        if (!"active".equalsIgnoreCase(status)) {
            System.out.println("-> [LoginService] Authentication failed: Account status is inactive (" + status + ") for user: " + username);
            throw new AccountStatusException("Account is restricted: " + status, status);
        }

        List<PartyRole> roles = partyRoleRepository.findByPersonPersonId(person.getPersonId());

        if (roles.isEmpty()) {
            return null;
        }

        PartyRole role = roles.get(0);

        return partyRoleMapper.toDto(role);
    }
}

File: src/main/kotlin/com/itsci/mju/maebanjumpen/login/service/impl/LoginServiceImpl.kt
package com.itsci.mju.maebanjumpen.login.service.impl

import com.itsci.mju.maebanjumpen.entity.*
import com.itsci.mju.maebanjumpen.login.dto.LoginDTO
import com.itsci.mju.maebanjumpen.login.repository.LoginRepository
import com.itsci.mju.maebanjumpen.login.service.LoginService
import com.itsci.mju.maebanjumpen.partyrole.dto.*
import com.itsci.mju.maebanjumpen.partyrole.repository.PartyRoleRepository
import com.itsci.mju.maebanjumpen.person.repository.PersonRepository
import org.springframework.stereotype.Service

import org.springframework.transaction.annotation.Transactional

@Service
class LoginServiceImpl(
    private val loginRepository: LoginRepository,
    private val personRepository: PersonRepository,
    private val partyRoleRepository: PartyRoleRepository,
) : LoginService {

    private fun mapLoginToDto(login: Login): LoginDTO {
        return LoginDTO(
            username = login.username,
            password = null
        )
    }

    private fun mapPartyRoleToDto(partyRole: PartyRole): PartyRoleDTO {
        return when (partyRole) {
            is Hirer -> HirerDTO().apply {
                id = partyRole.id?.toInt()
                balance = partyRole.balance
            }
            is Housekeeper -> HousekeeperDTO().apply {
                id = partyRole.id?.toInt()
                balance = partyRole.balance
                photoVerifyUrl = partyRole.photoVerifyUrl
                statusVerify = partyRole.statusVerify?.name
                rating = partyRole.rating
                dailyRate = partyRole.dailyRate
            }
            is Admin -> AdminDTO().apply {
                id = partyRole.id?.toInt()
                adminStatus = partyRole.adminStatus
            }
            is AccountManager -> AccountManagerDTO().apply {
                id = partyRole.id?.toInt()
                managerID = partyRole.managerID?.toInt()
            }
            is Member -> MemberDTO().apply {
                id = partyRole.id?.toInt()
                balance = partyRole.balance
            }
            else -> MemberDTO().apply {
                id = partyRole.id?.toInt()
            }
        }
    }

    @Transactional(readOnly = true)
    override fun authenticate(username: String, password: String): PartyRoleDTO? {
        return findPartyRoleByLogin(username, password)
    }

    @Transactional
    override fun CreateLogin(loginDto: LoginDTO): LoginDTO {
        requireNotNull(loginDto.username) { "Username cannot be null" }
        requireNotNull(loginDto.password) { "Password cannot be null" }

        val login = Login(
            username = loginDto.username!!,
            password = PasswordUtil.hashPassword(loginDto.password!!)
        )
        val savedLogin = loginRepository.save(login)
        return mapLoginToDto(savedLogin)
    }

    @Transactional(readOnly = true)
    override fun getLoginByUsername(username: String): LoginDTO? {
        return loginRepository.findById(username)
            .map { mapLoginToDto(it) }
            .orElse(null)
    }

    @Transactional
    override fun deleteLogin(username: String) {
        loginRepository.deleteById(username)
    }

    override fun updateLogin(username: String, loginDto: LoginDTO): LoginDTO? {
        return null
    }

    @Transactional(readOnly = true)
    override fun findPartyRoleByLogin(username: String, rawPassword: String): PartyRoleDTO? {
        val loginOpt = loginRepository.findByUsername(username)
        if (loginOpt.isEmpty) {
            return null
        }

        val storedLogin = loginOpt.get()
        val storedHash = storedLogin.password

        if (!PasswordUtil.verifyPassword(rawPassword, storedHash)) {
            println("-> [LoginService] Authentication failed: Invalid password for user: $username")
            return null
        }

        val personOpt = personRepository.findByLoginUsername(username)
        if (personOpt.isEmpty) {
            println("-> [LoginService] Authentication failed: Person not found for user: $username")
            return null
        }
        val person = personOpt.get()

        val status = person.accountStatus
        if (!status.equals("active", ignoreCase = true)) {
            println("-> [LoginService] Authentication failed: Account status is inactive ($status) for user: $username")
            throw BadRequestExceptßion("Account is restricted: $status")
        }

        val roles = partyRoleRepository.findByPersonId(person.id!!)

        if (roles.isEmpty()) {
            return null
        }

        val role = roles[0]
        return mapPartyRoleToDto(role)
    }
}

